docs/architecture/pk-4.md

```diff
@@ -0,0 +1,400 @@
+* Encrypt/decrypt: `mbedtls_pk_decrypt()`, `mbedtls_pk_encrypt()`. Use PSA.
+
+### Design philosophy for the new PK
+
+We retain the concept of a “PK context”, which can either be empty or contain a public key or contain a key pair. The new PK handles key parsing and writing, and has convenience functions to sign with a key.
+
+A PK context has the following conceptual properties:
+
+* A PSA key type (key pair or public key). This is `PSA_KEY_TYPE_NONE` for an empty context.
+* A PSA algorithm used for signing and verification. This is `PSA_ALG_NONE` for an empty context. When a key is set, this is a signature algorithm or a signature wildcard algorithm policy (which specifies a signature mechanism, but leaves a hash algorithm unspecified).
+* Key material that matches the key type. This can be directly in the PK object, or indirectly via a PSA key identifier.
+* Optionally, an associated PSA key identifier. The PSA key may be owned by the PK context and destroyed when the context is destroyed, or it may be referenced by the PK context and left alone when the context is destroyed.
+
+The PK module does not enforce key policies. In particular, it is possible to copy a PK context into a context with a different signature algorithm.
+
+### Private interfaces
+
+In this document, a ***private*** interface is one that is not documented. Applications should not use private interfaces, and we do not promise any kind of stability about them. Mbed TLS can use private interfaces of TF-PSA-Crypto, but in the medium term (over the lifetime of TF-PSA-Crypto 1.x and Mbed TLS 4.x), it should stop doing so. Public interfaces must not rely on private interfaces, for example a private type cannot be used in the prototype of a public function. However, public types can have a private implementation (we guarantee that the type will keep existing, but it may be implemented differently, typically adding and removing fields in a structure).
+
+An ***internal*** interface is only usable inside TF-PSA-Crypto.
+
+## API elements
+
+### PK context type
+
+#### Keep `mbedtls_pk_context`
+
+We keep `mbedtls_pk_context` largely as it is now. It will be reworked after 1.0 as needed. Keep:
+
+```
+mbedtls_pk_context
+mbedtls_pk_init()
+mbedtls_pk_free()
+```
+
+#### Context metadata
+
+We explicitly associate an algorithm with each PK context. This is intended to match the algorithm in the underlying PSA key when there is one. Having an algorithm stored in the context lets us treat PK contexts more uniformly regardless of whether they contain a PSA key or a direct pointer to key material.
+
+ACTION (https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/204): populate the field `pk->psa_algorithm` when populating `pk`. Use the same approach as the existing function `mbedtls_pk_get_psa_attributes()` for a signature usage.
+
+ACTION (https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/204): implement a new function to change the algorithm associated with a PK context:
```

Comment:
Having thought about it further, this is definitely wrong. It can't be implemented in a reasonable way on top of a PSA key, since you can't change the PSA key's policy, and having a PK algorithm that doesn't match the policy of the underlying PSA key is pointless.

We need a different solution for the following scenario:

1. I parse a key.
2. Based on some of the key's metadata, I decide which signature algorithm to use it with.
3. I sign with pk.

There is already such a workflow. It's somewhat inefficient, but I think we can live with it.

1. Parse the key.
2. `mbedtls_pk_get_psa_attributes()` and `mbedtls_pk_import_into_psa()`. This workflow was created precisely to solve this problem.
3. `mbedtls_pk_copy_from_psa()` or, to save some resources, `mbedtls_pk_wrap_psa()`.